src/Compilers/CSharp/Test/Semantic/Semantics/StructConstructorTests.cs

```diff
@@ -0,0 +1,1669 @@
+  IL_0097:  box        ""S3""
+  IL_009c:  call       ""void System.Console.WriteLine(object)""
+  IL_00a1:  ret
+}");
+            verifier.VerifyIL("S1..ctor()",
+@"{
+  // Code size       13 (0xd)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  ldc.i4.1
+  IL_0002:  box        ""int""
+  IL_0007:  stfld      ""object S1.<X>k__BackingField""
+  IL_000c:  ret
+}");
+            verifier.VerifyIL("S2..ctor()",
+@"{
+  // Code size       25 (0x19)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  ldc.i4.2
+  IL_0002:  box        ""int""
+  IL_0007:  stfld      ""object S2.<X>k__BackingField""
+  IL_000c:  ldarg.0
+  IL_000d:  ldc.i4.2
+  IL_000e:  box        ""int""
+  IL_0013:  call       ""void S2.Y.init""
+  IL_0018:  ret
+}");
+            verifier.VerifyMissing("S3..ctor()");
+            verifier.VerifyIL("S3..ctor(object)",
+@"{
+  // Code size       25 (0x19)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  ldc.i4.3
+  IL_0002:  box        ""int""
+  IL_0007:  stfld      ""object S3.<X>k__BackingField""
+  IL_000c:  ldarg.0
+  IL_000d:  ldc.i4.3
+  IL_000e:  box        ""int""
+  IL_0013:  call       ""void S3.Y.set""
+  IL_0018:  ret
+}");
+        }
+
+        [Fact]
+        public void FieldInitializers_04()
+        {
+            var source =
+@"#pragma warning disable 649
+using System;
+struct S1<T> { internal int X = 1; }
+struct S2<T> { internal int X = 2; public S2() { } }
+struct S3<T> { internal int X = 3; public S3(int _) { } }
+class Program
+{
+    static void Main()
+    {
+        Console.WriteLine(new S1<object>().X);
+        Console.WriteLine(new S2<object>().X);
+        Console.WriteLine(new S3<object>().X);
+    }
+}";
+
+            var verifier = CompileAndVerify(source, parseOptions: TestOptions.RegularPreview, options: TestOptions.ReleaseExe, expectedOutput:
+@"1
+2
+0");
+            verifier.VerifyIL("Program.Main",
+@"{
+  // Code size       50 (0x32)
+  .maxstack  1
+  .locals init (S3<object> V_0)
+  IL_0000:  newobj     ""S1<object>..ctor()""
+  IL_0005:  ldfld      ""int S1<object>.X""
+  IL_000a:  call       ""void System.Console.WriteLine(int)""
+  IL_000f:  newobj     ""S2<object>..ctor()""
+  IL_0014:  ldfld      ""int S2<object>.X""
+  IL_0019:  call       ""void System.Console.WriteLine(int)""
+  IL_001e:  ldloca.s   V_0
+  IL_0020:  initobj    ""S3<object>""
+  IL_0026:  ldloc.0
+  IL_0027:  ldfld      ""int S3<object>.X""
+  IL_002c:  call       ""void System.Console.WriteLine(int)""
+  IL_0031:  ret
+}");
+        }
+
+        [Fact]
+        public void FieldInitializers_05()
+        {
+            var source =
+@"#pragma warning disable 649
+using System;
+class A<T>
+{
+    internal struct S1 { internal int X { get; } = 1; }
+    internal struct S2 { internal int X { get; init; } = 2; public S2() { } }
+    internal struct S3 { internal int X { get; set; } = 3; public S3(int _) { } }
+}
+class Program
+{
+    static void Main()
+    {
+        Console.WriteLine(new A<object>.S1().X);
```

Comment:
>S1 [](http://example.com/codeflow?start=40&length=2)

Consider verifying `HasFieldInitializers()` on these types. I'd expect that with the current implementation, we'd see `false` on the `SubstitutedNestedNamedTypeSymbol` invovled since there is no override of the default `virtual` base.

Reply:
> I'd expect that with the current implementation, we'd see `false` on the `SubstitutedNestedNamedTypeSymbol` ...

`HasFieldInitializers()` should only be called on a definition. Added an assert to that method to make it clear.